tiny_dnn/core/framework/tensor.h

```diff
@@ -148,58 +150,57 @@ class Tensor {
 #endif
 
     // Returns the tensor shape
-    const std::array<size_t, 4>& shape() const { return shape_; }
+    const std::array<size_t, kDimensions>& shape() const { return shape_; }
 
     // Returns the value of a specified index in the tensor.
     // Checked version (throw exceptions for out-of-range error)
-    U& host_at(const size_t d0,
-               const size_t d1,
-               const size_t d2,
-               const size_t d3) {
-        return *host_ptr(d0, d1, d2, d3);
+    template<typename... Args>
+    U& host_at(const Args... args) {
+        return *host_ptr(args...);
+    }
+    template<typename... Args>
+    U host_at(const Args... args) const {
+        return *host_ptr(args...);
     }
 
-    U host_at(const size_t d0,
-              const size_t d1,
-              const size_t d2,
-              const size_t d3) const {
-        return *host_ptr(d0, d1, d2, d3);
+    size_t host_pos(const size_t d) {
+        if (d >= *shape_.rbegin())  {
+            throw nn_error("Access tensor out of range.");
+        }
+        return d;
     }
 
-    // Returns the pointer to a specified index in the tensor
-    // Checked version (throw exceptions for out-of-range error)
-    const U* host_ptr(const size_t d0,
-                      const size_t d1,
-                      const size_t d2,
-                      const size_t d3) const {
-        if (d0 >= shape_[0] || d1 >= shape_[1] ||
-            d2 >= shape_[2] || d3 >= shape_[3]) {
+    template<typename... Args>
+    size_t host_pos(const size_t d,
+                const Args... args) {
+        size_t dim = shape_.size() - sizeof...(args) - 1;
+        if (d >= shape_[dim])  {
             throw nn_error("Access tensor out of range.");
         }
+        size_t shift = 1;
+        for (size_t i = dim + 1; i < shape_.size(); ++i)
+            shift *= shape_[i]; //TODO(Randl): optimize
 
-        return host_data() + (
-            shape_[1] * shape_[2] * shape_[3] * d0 +
-            shape_[1] * shape_[2] * d3 +
-            shape_[1] * d2 +
-            d1
-            );
+        return (d * shift + host_pos(args...) );
     }
 
-    U* host_ptr(const size_t d0,
-                const size_t d1,
-                const size_t d2,
-                const size_t d3) {
-        if (d0 >= shape_[0] || d1 >= shape_[1] ||
-            d2 >= shape_[2] || d3 >= shape_[3]) {
-            throw nn_error("Access tensor out of range.");
+    // Returns the pointer to a specified index in the tensor
+    // Checked version (throw exceptions for out-of-range error)
+    template<typename... Args>
+    const U* host_ptr (const Args... args) const {
+        //  TODO(Randl): Should we throw here if lower number of arguments?
+        if (sizeof...(args) != shape_.size())  {
```

Comment:
If I understand correctly, the number of variadic template parameters in here should be equal to kDimensions.

Using `static_assert` does not incur runtime cost but then `host_ptr` method will be able to throw an exception.

```
static_assert(sizeof...(args) == kDimensions, "number of parameters must be... y'know");
static_assert(check<std::is_convertible, int, Args...>, "!");
```